website/blog/2023-07-14-Local-LLMs/index.mdx

```diff
@@ -0,0 +1,107 @@
+---
+title: Use flaml.autogen for local LLMs
+authors: jialeliu
+tags: [LLM, FLAMLv2]
+---
+**TL;DR:**
+We demonstrate how to use flaml.autogen for local LLM application. As an example, we will initiate an endpoint using [FastChat](https://github.com/lm-sys/FastChat) and perform inference on [ChatGLMv2-6b](https://github.com/THUDM/ChatGLM2-6B).
+
+## Preparations
+
+### Clone FastChat
+
+FastChat provides OpenAI-compatible APIs for its supported models, so you can use FastChat as a local drop-in replacement for OpenAI APIs. However, it's code needs minor modification in order to function properly.
+
+```bash
+git clone https://github.com/lm-sys/FastChat.git
+cd FastChat
+```
+
+According to this [issue](https://github.com/lm-sys/FastChat/issues/1641), comment out all the line containing `finish_reason` in `fastchat/protocol/api_protocal.py` and `fastchat/protocol/openai_api_protocol.py` will fix the errors. This is what modified code looks like:
+
+```python
+class CompletionResponseChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[int] = None
+    # finish_reason: Optional[Literal["stop", "length"]]
+
+class CompletionResponseStreamChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[float] = None
+    # finish_reason: Optional[Literal["stop", "length"]] = None
+```
+
+### Download checkpoint
+
+ChatGLM-6B is an open bilingual language model based on General Language Model (GLM) framework, with 6.2 billion parameters. ChatGLM2-6B is its second-generation version.
+
+Before downloading from HuggingFace Hub, you need to have Git LFS [installed](https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage).
+
+```bash
+git clone https://huggingface.co/THUDM/chatglm2-6b
+```
+
+## Initiate server
```

Comment:
Great job! I have a suggestion: we may want to add a bash script that inputs the model path and builds the endpoint for user. It would be very nice to have this bash cloned a modified version of FastChat to reduce user effort.

Reply:
Yes! This is on my to-do list. I will try to include it in the next update.

Reply:
no I didn't

Reply:
I wonder if it's because of pydantic version. The error looks related to type check by pydantic.

Reply:
@LeoLjl given that @JieyuZ2 has no encountered this issue, how about adding a note "if you encounter this error: ..." so that users know when to modify the code?

Reply:
> @LeoLjl given that @JieyuZ2 has no encountered this issue, how about adding a note "if you encounter this error: ..." so that users know when to modify the code?

Of course.